### ISSUE TITLE
Feature request: alt text for cover images

### ISSUE THREAD

**Issue by fluffynuts** [forem/forem#4758]
As per the title -- please can someone implement alt-text for cover images? I'd really like it if people with vision difficulties could also understand what's going on in my cover images. Something like:

cover_image_alt: Book-case in a library

In the meta section of the post would be really helpful.

**Comment by rhymes** [comment]
Hi @fluffynuts, this makes total sense! 

We just need to understand what's the best strategy here.

Currently:

- as @fluffynuts pointed out, the cover image don't have an alternative text
- the cover image is rendered as a background image on a `<div>`:

```html
<section>
          <div class="image image-final" style="background-color:#dddddd;background-image:url(https://res.cloudinary.com/practicaldev/image/fetch/s--OFcn6FWy--/c_imagga_scale,f_auto,fl_progressive,h_420,q_auto,w_1000/https://res.cloudinary.com/practicaldev/image/fetch/s--hhp5MS6x--/c_imagga_scale%2Cf_auto%2Cfl_progressive%2Ch_420%2Cq_auto%2Cw_1000/https://thepracticaldev.s3.amazonaws.com/i/o1ob2lhokyymj5fikoxl.jpg)">
          </div>
      </section>
```

from https://dev.to/giteden/how-to-gradually-build-a-design-system-2g3b

Possible solution:

1. switch the cover to an `<img alt="">` tag or use an ARIA label (though this option might be less semantically relevant, especially to screen readers)
1. have a default alt text, for example "cover image"
1. add `cover_image_alt_text` to the DB
1. add `cover_image_alt_text` to the front matter for editor v1
1. add a field to fill in the `cover_image_alt_text` in the image upload editor

Anything else? Other possible solutions?

What do you think @fluffynuts ?

**Comment by fluffynuts** [comment]
Hey, thanks for asking (:

After a little thought, I would probably like 1 & 2 & 3 & 4 😀
- having the cover image presented as an image instead of a div should surely be better for screeen-readers and simplifies the alt-text implementation.
- 3 & 4 would (I imagine) be required to implement the feature, 
- and it would be a Good Idea to have some default alt text (2), again to help out screen readers.

5 becomes a problem in that the image uploader is used to upload _any_ image -- and inline images already have a markdown method for displaying the alt-text (and I like how the semantics for images in markdown kind of encourage the alt-text)

The combination of 1-4 (including 2 for the default) also makes this really easy to opt-in for and solves the issue where the author doesn't opt-in, but we still want to provide the best experience possible for visually-challenged readers.

So that's my 2c (:

**Comment by aitchiss** [comment]
I'm so pleased to see this feature request already exists! I have been thinking the same. We should avoid option 2 though - redundant alt text is a common accessibility failing, and unless we are telling users what the image is in the `alt` then it's not helpful (if you use an `img` element, then the screen reader will already tell them it's an image)

**Comment by rhymes** [comment]
> redundant alt text is a common accessibility failing, and unless we are telling users what the image is in the alt then it's not helpful (if you use an img element, then the screen reader will already tell them it's an image)

yeah, you're right, it's the same with regular images, no default is the same with a default that says "image"

**Comment by fluffynuts** [comment]
The only reason I suggested it is to disambiguate: a cover image isn't really important for the story (well, normally), so, if I couldn't see it, being able to know that it's not just some random image, but something put out there just to make the article look nice, would allow me to move past it.

I must, however, stress that, whilst I'm NAT, I'm not visually impaired (not to a point that matters -- minor visual snow, discomfort with bright things), so my input may not be valuable on this point.

**Comment by mattpeet** [comment]
> The only reason I suggested it is to disambiguate: a cover image isn't really important for the story (well, normally), so, if I couldn't see it, being able to know that it's not just some random image, but something put out there just to make the article look nice, would allow me to move past it.

That's a good point, and totally makes sense. The standard practice for purely decorative images is to have an empty alt `alt=""` (some notes about it here: https://webaim.org/techniques/alttext/#decorative) - so maybe that could be the way to go for a default option?

**Comment by aitchiss** [comment]
> The only reason I suggested it is to disambiguate: a cover image isn't really important for the story (well, normally), so, if I couldn't see it, being able to know that it's not just some random image, but something put out there just to make the article look nice, would allow me to move past it.

That's a good point, and totally makes sense. The standard practice for purely decorative images is to have an empty alt `alt=""` (some notes about it here: https://webaim.org/techniques/alttext/#decorative) - so maybe that could be the way to go for a default option?

**Comment by fluffynuts** [comment]
@aitchiss I'm not an expert in accessibility, so if that's what works best, I'm all for it (:

**Comment by moopet** [comment]
I think this isn't necessary for pure cosmetics.
I do, however, think that if an image is semantically part of the post, even if it's the cover and purely for decoration, it should be rendered as an image or picture element and given an empty alt attribute by default. We could add `cover_image_alt` to the front matter to allow people to override the empty attribute if they believe that the image is important to the post.

**Comment by ooloth** [comment]
Once implemented, it would be great if `cover_image_alt` could be included in the article data available via the API as well.

**Comment by nickytonline** [comment]
The cover image is actually a background image of an `< a href />` which has a title attribute and there is also text for the link now.

![image](https://user-images.githubusercontent.com/833231/94726645-84b71480-032b-11eb-9b37-87ad264c4176.png)

@ooloth, the API feature request seems like a separate issue.

**Comment by cmgorton** [comment]
@nickytonline given the new way this is rendered with a title attribute do you think this issue can be closed or should we consider some of the other option proposed in this issue?

**Comment by nickytonline** [comment]
I think it's probably OK to close this because as mentioned, the API feature request is something separate and there is no `<img />` tag now.

It seems to be fine in terms of accessibility for the issue mentioned, but there are adjacent identical links which are showing up as alerts in WAVE. That could be something to look into in a separate issue.

![image](https://user-images.githubusercontent.com/833231/105533858-b1b47c00-5cba-11eb-9f93-953eadffa98d.png)
